R/type-list-of.R

```diff
@@ -212,6 +227,19 @@ vec_cast.vctrs_list_of.vctrs_list_of <- function(x, to, ...) {
   list_as_list_of(x, ptype = ptype)
 }
 
+#' @export
+vec_cast.list.vctrs_list_of <-function(x, to, ...) {
+  vec_data(x)
+}
+#' @export
+vec_cast.vctrs_list_of.list <-function(x, to, ...) {
+  list_as_list_of(
+    x,
+    attr(to, "ptype"),
+    error_call = caller_env()
```

Comment:
I see how if you remove this you get `vec_cast.vctrs_list_of.list()` in the error:

```r
> vec_cast(list(1), list_of("x"))
Error in `vec_cast.vctrs_list_of.list()`:
! Can't convert `..1` <double> to <character>.
Run `rlang::last_trace()` to see where the error occurred.
```

But I'd really rather not pay for calling `caller_env()` every time, especially because it doesn't give you the right call

``` r
library(vctrs)

# Don't see `fn()`
fn <- function() vec_cast(list(1), list_of("x"))
fn()
#> Error:
#> ! Can't convert `..1` <double> to <character>.

fn <- function() vec_cast(1, "x")
fn()
#> Error in `fn()`:
#> ! Can't convert `1` <double> to <character>.
```

Could we do this since `call` is in the generic? It seems to give the right call and shouldn't be evaluated unless it is required.

```r
vec_cast.vctrs_list_of.list <-function(x, to, ..., call = caller_env()) {
  list_as_list_of(
    x,
    attr(to, "ptype"),
    error_call = call
  )
}
```

Reply:
We are not paying for it every time, it's evaluated lazily since it's passed as argument.

Good point regarding the generic argument. I'll take a look and make sure we pass `call` through `...` down to methods.